File: src/org/parboiled/ParsingResult.java
package org.parboiled;

import org.jetbrains.annotations.NotNull;
import org.parboiled.support.InputBuffer;
import org.parboiled.support.ParseError;

import java.util.List;

/**
 * A simple container for encapsulating the result of a parsing run.
 */
public class ParsingResult {

    /**
     * The root node of the parse tree created by the parsing run.
     */
    public final Node root;

    /**
     * The list of parse errors created during the parsing run.
     */
    public final List<ParseError> parseErrors;

    /**
     * The underlying input buffer.
     */
    public final InputBuffer inputBuffer;

    ParsingResult(Node root, @NotNull List<ParseError> parseErrors, InputBuffer inputBuffer) {
        this.root = root;
        this.parseErrors = parseErrors;
        this.inputBuffer = inputBuffer;
    }

    /**
     * @return true if the parsing run was completed without errors.
     */
    public boolean hasErrors() {
        return !parseErrors.isEmpty();
    }

}

File: src/main/scala/org/parboiled/scala/ParsingResult.scala
package org.parboiled.scala

import org.parboiled.Node
import org.parboiled.support.{InputBuffer, ValueStack}
import org.parboiled.errors.ParseError
import org.parboiled.support.{ParsingResult => PParsingResult}
import annotation.unchecked.uncheckedVariance

object ParsingResult {
  def apply[V](result: PParsingResult[V])  = new ParsingResult[V](result)

  implicit def unwrap[V](result: ParsingResult[V]): PParsingResult[V] = result.inner
}

class ParsingResult[+V](val inner: PParsingResult[V] @uncheckedVariance) {
  val matched: Boolean = inner.matched
  val resultValue: V = inner.resultValue
  val result: Option[V] = if (inner.resultValue != null) Some(inner.resultValue) else None
  val parseErrors: List[ParseError] = List(inner.parseErrors.toArray(new Array[ParseError](inner.parseErrors.size)): _*)
  val parseTreeRoot: Node[V] @uncheckedVariance = inner.parseTreeRoot
  val valueStack: ValueStack[Any] = inner.valueStack.asInstanceOf[ValueStack[Any]]
  val inputBuffer: InputBuffer = inner.inputBuffer

  def hasErrors = !parseErrors.isEmpty

}